### PR TITLE
Support for fully contained iterative dns resolutions (no name server necessary)

### PR DESCRIPTION
**Pull request by paul-pearce**
This code, much like the rest of zdns, is in a beta state. Comments and further testing are welcome.

### CODE REVIEW

File: .travis.yml
Comment (dadrian) [inline .travis.yml:3]: Should maybe also build against 1.7.x and 1.8.x?

File: integration_tests.py
Comment (dadrian) [inline integration_tests.py:233]: What are these tests in Python and not in Go?

File: interface.go
Comment (dadrian) [inline interface.go:88]: Add a comment explaining this int

Comment (zakird) [inline interface.go:88]: what is this int??? This seems sketchy to me.

Comment (dadrian) [reply interface.go:88]: Please run go lint on your code and fix some of the style errors, for example write comments following convention.

File: lookup.go
Comment (dadrian) [inline lookup.go:69]: I don't think you need to deference this to call a method on it

File: modules/axfr/axfr.go
Comment (zakird) [inline modules/axfr/axfr.go:150]: This really seems like something that should be available as a global or something and not something that we have to be passing around. I remain _very_ skeptical about this.

File: zdns/main.go
Comment (dadrian) [inline zdns/main.go:96]: s/recursive/iterative

File: zdns/main.go
Comment (dadrian) [inline zdns/main.go]: I don't see why we need `--timeout` and `--iterative-timeout`. Why are these not just the same?

File: modules/nslookup/nslookup.go
Comment (zakird) [inline modules/nslookup/nslookup.go]: Wait, why is this defined here and not in `miekg.RoutineLookupFactory`?

File: integration_tests.py
Comment (dadrian) [reply integration_tests.py:233]: Ok so, I understand writing some stuff in Python to actually call the binary we build, but why are there _no_ tests in Go?

**Review by dadrian — DISMISSED** [review]

File: zdns/main.go
Comment (zakird) [inline zdns/main.go]: How is this different than normal timeout?

**Comment by zakird** [comment]
I don't really like how this threadID is getting passed around. @dadrian thinks we should pass around the name we were originally trying to lookup. I think we should use something like http://blog.sgmansfield.com/2015/12/goroutine-ids/.

File: interface.go
Comment (paul-pearce) [reply interface.go:88]: Deferring this to the threadID discussion below.

File: modules/axfr/axfr.go
Comment (paul-pearce) [reply modules/axfr/axfr.go:150]: Deferring this to the threadID discussion below

File: modules/nslookup/nslookup.go
Comment (paul-pearce) [reply modules/nslookup/nslookup.go]: Good catch. Removed.

File: .travis.yml
Comment (paul-pearce) [reply .travis.yml:3]: I'll defer to you all on this decision. I simply rev'd to the version that's common across both our infrastructures.

File: integration_tests.py
Comment (paul-pearce) [reply integration_tests.py:233]: Deferring to Zakir as he setup the testing infrastructure.

File: interface.go
Comment (paul-pearce) [reply interface.go:88]: The head of master has 220 golint errors. This branch reduces the number of golint errors down to 150. I suggest we tackle conforming to golint as a separate PR across the entire project.

File: lookup.go
Comment (paul-pearce) [reply lookup.go:69]: Is your thinking?

`f, err := g.MakeRoutineFactory(threadID)`

If so:
`
../lookup.go:69: g.MakeRoutineFactory undefined (type *GlobalLookupFactory is pointer to interface, not interface)
`

File: zdns/main.go
Comment (paul-pearce) [reply zdns/main.go]: Hmm. iterativeTimeout was intended to be set to 10 by default, to match bind functionality. That's my mistake.

As for why they are different: They are for fundamentally different functions. timeout's primary job is to govern interacting with BIND instances that have a specific timeout default (generally 10 seconds), so we set it >10. iterativeTimeout is more governing an iterative process that involves interacting with multiple servers.

I'd suggest we keep the 3 separate to allow us to, by default, mimic the BIND timeout behavior.

Comment (paul-pearce) [reply zdns/main.go]: See discussion above.

**Comment by paul-pearce** [comment]
WRT ThreadID. I preferred the solution of passing the the ID context to constructors more than using something that begins "if you use this, you will go straight to hell."

@zakird can you clarify why you dislike passing the context to the constructors?

File: interface.go
Comment (dadrian) [reply interface.go:88]: My point was to just try to hit some of the low hanging fruit, such as comment style because they're easy, not to imply you had to fix all the go lint errors.

**Review by dadrian — APPROVED** [review]
I'd like to see issues filed for:

1. Refactoring this to have some sort of context that's passed around in a sane way. I don't have a good sense for how the code is laid out currently, but given that it's so damn hard to track Goroutines, my sense is that something should be architected differently.

2. More things should be tested in Golang itself

3. We should clean up all the comments, so that the comments provide documentation suitable for godoc.org.

4. It seems like we're doing some non-Go-things, like passing pointers-to-interfaces around, instead of interfaces. We should clean those up.

But none of these are @paul-pearce's fault, at least within the scope of iterative resolution. I am just worried about ZDNS turning into a ZGrab project of Go debt.

I'll approve and leave the rest of the review to @zakird

**Comment by zakird** [comment]
Given that @dadrian's comments are non-specific and largely unrelated to this branch, I'm mostly fine putting those aside for now. However, I still do not understand these timeout flags (even after reading @paul-pearce's response several times), and I don't want to merge until we have more helpful descriptions.

**Comment by zakird** [comment]
@paul-pearce my concern with the threadID is that it's really ugly to have to pass an integer to _every_ function call in the entire application for debugging purposes.

**Comment by zakird** [comment]
@paul-pearce when you're online, can you ping me, and I can try to better understand what these command line flags do.

**Review by zakird — CHANGES_REQUESTED** [review]
Blocking until we can sort out command line flags.

File: integration_tests.py
Comment (zakird) [reply integration_tests.py:233]: While I agree that unit tests would be helpful in this code base, these provide coverage over core functionality, and there isn't any benefit of writing a test handler in go here instead of python. Go-based tests are not inherently better.

**Comment by paul-pearce** [comment]
@zakird, re "ThreadID being in _every_ function call". I don't follow. It is only passed into the MakeRoutineFactory() constructors for each module. There's one logging function in miekg that it's also passed into, since it's called from a global context instead of a routine context. Is there confusion on the scope of the state, or language issues?

RE command line arguments. As discussed outside this thread, I'd personally like to see the following defaults:

timeout -> 15s, to interface with default bind timeout, without races
iterativeTimeout -> 10s, to match default bind
iterationTimeout -> 3s, to match default bind

I only broke the 3 out as separate options to emulate the desired 3 defaults (and then forget to reset them after testing, sigh).

But honestly I don't care enough to hold up the merge. If there's broad agreement, I'm fine with:

timeout -> 15s, used for both iterative and recursive timeouts
iterationTimeout -> 3 or 4s.

**Comment by zakird** [comment]
I have fixed timeout. 

re "ThreadID being in every function call" => things like `GetCachedResult` now have to take a threadID. That's what I find incredibly ugly.

**Review by zakird — APPROVED** [review]

**Pull request merged**
